### DISCUSSION TITLE
Random canvas creation not working as predicted or described

### DISCUSSION THREAD

**Discussion by Traitorsclaw** [ImageMagick/ImageMagick#1943] [General] [answered]
I'm far from an expert in ImageMagick, so I decided to have a bit of a play recently with basic canvas creation ... specifically by trying some of the examples here: [https://www.imagemagick.org/Usage/backgrounds
](url)
What I've found is that most of these examples seem to work pretty much as predicted/described, however, for me, some of the transformations are NOT working as described. I'm not sure if this is due to outdated documentation in that thread (ie. maybe IM functionality has changed since that thread was written), a syntactic issue with my environment, or simply my lack of understanding.
  
My environment is WIN7, ImageMagick-6.7.8-Q16 (64-bit)

I'd appreciate any advice/comments.

I wrote a very simple batch script which creates the base images, and allows me to provide any particular transform as a quoted parameter, and produce the 3 outputs (plasma_fractal.png, random_noise.png, and random_hex_tile.png), with the selected transformation applied... 
Here's the code ... 

> @ECHO OFF
set transform=%1
if (%transform%)==() (
	set transform=""
)
set transform=%transform:"=%
set noauto=$2
convert -seed 50 -size 120x120  plasma:fractal temp.png
convert temp.png %transform% -auto-level -shave 20x20 +repage plasma_fractal.png
 convert -seed 50 -size 80x80 xc: +noise Random temp.png
convert temp.png -virtual-pixel tile %transform% -auto-level random_noise.png
convert -seed 50 -size 50x80 xc: +noise Random -write mpr:rand -extent 100x80 -page +50-40 mpr:rand -page +50+40 mpr:rand -flatten temp.png
convert temp.png -virtual-pixel tile %transform% -auto-level random_hex_tile.png
del temp.png

eg. At the command line...

> test
will produce the base images:
![plasma_fractal](https://user-images.githubusercontent.com/25820859/80336019-2bddde80-88aa-11ea-9bee-b5e7bc6d20b0.png) ![random_noise](https://user-images.githubusercontent.com/25820859/80336023-2ed8cf00-88aa-11ea-8eae-e0e486491e86.png) ![random_hex_tile](https://user-images.githubusercontent.com/25820859/80336021-2da7a200-88aa-11ea-99d2-def34569e7f4.png) 

> test "-blur 0x2  -edge 10"
gives:
![plasma_fractal](https://user-images.githubusercontent.com/25820859/80336141-8d9e4880-88aa-11ea-85b7-1358187f8e3d.png) ![random_noise](https://user-images.githubusercontent.com/25820859/80336138-8bd48500-88aa-11ea-9f2c-c268add33e84.png) ![random_hex_tile](https://user-images.githubusercontent.com/25820859/80336139-8d05b200-88aa-11ea-934c-d90be012eaaf.png) 

However, I run into trouble with some of the others.
For example "pits" ... 
The "Background Examples" thread shows these outputs:
![fractal_pits](https://user-images.githubusercontent.com/25820859/80336298-07cecd00-88ab-11ea-89e5-8e472b5df128.png) ![noise_pits](https://user-images.githubusercontent.com/25820859/80336297-07363680-88ab-11ea-942c-e3df4546b7e0.png) ![hextile_ridges](https://user-images.githubusercontent.com/25820859/80336294-06050980-88ab-11ea-989e-5e1c52ecd0bb.png) 
... whereas ...

> test "-blur 0x10 -edge 1 -fx G -shade 280x45"
on my system, gives these results:
![plasma_fractal](https://user-images.githubusercontent.com/25820859/80336378-364ca800-88ab-11ea-815b-80272d6f990e.png) ![random_noise](https://user-images.githubusercontent.com/25820859/80336383-38166b80-88ab-11ea-9430-8954fc17fda0.png) ![random_hex_tile](https://user-images.githubusercontent.com/25820859/80336381-36e53e80-88ab-11ea-828e-1ae20c6ba467.png) 

"mottled" produces the same:

> test "-blur 0x10 -write mpr:save -negate -edge 1 -negate -fx G ( mpr:save -edge 1 -fx G ) -shade 280x45 -average"
![plasma_fractal](https://user-images.githubusercontent.com/25820859/80336678-11a50000-88ac-11ea-8aa1-af1358de5d0b.png) ![random_noise](https://user-images.githubusercontent.com/25820859/80336676-1073d300-88ac-11ea-9c81-292039c93055.png) ![random_hex_tile](https://user-images.githubusercontent.com/25820859/80336679-123d9680-88ac-11ea-8ab7-1778e8c6a139.png)

Other's, such as  "paint_3D" seem to be fine..

> test "-blur 0x10 -paint 3 ( +clone -shade 120x45 ) +swap  -compose overlay -composite"
![plasma_fractal](https://user-images.githubusercontent.com/25820859/80337006-df47d280-88ac-11ea-8fb1-b4d9596db5bb.png) ![random_noise](https://user-images.githubusercontent.com/25820859/80337004-dd7e0f00-88ac-11ea-921e-4ce7bf4e8246.png) ![random_hex_tile](https://user-images.githubusercontent.com/25820859/80337008-dfe06900-88ac-11ea-97ec-c773b48a8003.png)

**Comment by snibgo** [comment]
Your IM v6.7.8 is very very old. I suggest you install a more recent version, preferably v7.

However, with v7.0.8-64, I have found problems with "+noise Random". For example:
`magick -seed 50 -size 80x80 xc: +noise Random rand_temp2.png`

![rand_temp2](https://user-images.githubusercontent.com/33812368/80366802-c8d85180-8881-11ea-8dc4-3bf935dae1c1.png)

The result has bands of non-randomness. I suppose this is a bug.

**Reply by Traitorsclaw** [reply]
Thanks, but perhaps it's not such a good idea right now ... as I suspect you are currently unable to reproduce **any** of these cases correctly as that version is failing right at step 1.
I don't actually have any actual need for those tiles which I found were not working (it was just something I noticed that didn't seem right)...
... and I have work which relies on the current version I have installed, so I'm inclined NOT to risk an upgrade at this point in time (even if it is outdated).

I'm pretty sure the thread with these examples pre-dates both versions, so perhaps there was a bug introduced in the version I'm using ( perhaps related to the "-edge" ).
It would however be interesting to see if v7 does a better job of these tiles with dlemstra's patch applied (or if the problem persists).

**Comment by dlemstra** [comment]
You are correct @snibgo. There is a repetition bug when OpenCL is enabled. Just pushed a patch to fix this.

**Comment by fmw42** [comment, marked as answer]
This command that you say does not work, works fine for me on IM 6.9.11.9 Q16 mac OSX

```
convert -size 120x120  plasma:fractal fractal.png
convert fractal.png -blur 0x10 -edge 1 -fx G -shade 280x45 -shave 20x20 +repage -auto-level result.png

```

![x](https://user-images.githubusercontent.com/13458218/80551828-a3c70a00-8979-11ea-987a-b6e23b95d053.png)

Perhaps you have a bug in your code.  What was your exact version of ImageMagick 6.7.8.x? If 6.7.8.9 that is known to have bugs. Perhaps you should upgrade.

**Reply by Traitorsclaw** [reply]
Thanks for the feedback.
Version: ImageMagick 6.7.8-5 2012-07-18 Q16

Although v7 may be a leap too far at the moment, it might be worth that small upgrade from 6.7 to 6.9 ... however it's also possible that this issue might be down to a WIN vs MAC difference ... for example, I'm yet to work out how to handle the single quotes (') and some of the other features of many IM commands ... which are problematic in the context of the windows CMD prompt (although that's not an issue in this particular very simple case).

**Reply by fmw42** [reply]
Use double quotes in Windows, in general, if you even need them. What command needs single quotes. IM 6.7.8.5 was right at the end of some significant changes to IM. I would recommend you upgrade at least to the current IM 6.9.11.9 or whatever is the closest you can get.  Did you try my command without all your extra code?

**Reply by Traitorsclaw** [reply]
Trouble with the quotes is probably due to the fact that I'm passing the *transformations* into a batch script as a quoted string, then unquoting it. ( Not a big deal ... just DOS issues! Grrr! )

Yes I did run it outside of the batch script ... same prob with the 6.7 version ... however, you'll see from my last reply that since then I have upgraded to a 6.9 version ... which fixes the problem (at least the "pits" one we focused on anyway.)

Thanks for that.

EDIT: Somehow, my *last reply" disappeared into the ether... (strange)
Here it is again:
![result](https://user-images.githubusercontent.com/25820859/80556916-48545500-8a29-11ea-9e8c-ef725612261d.png)

( 6.9.9-37 Q16 x64 )

**Reply by fmw42** [reply]
OK. But out of curiosity, if you upgraded, why not go for the latest version of 6.9.11.9?

**Reply by Traitorsclaw** [reply]
Wasn't listed here: [https://sourceforge.net/projects/imagemagick/files/im6-exes/](url)

Is there a [ better / more recent ] official download page for these ?

**Reply by Traitorsclaw** [reply]
> 
> 
> What command needs single quotes.

Most of the "gradient transforms" ....

... which is an issue for DOS batch scripts because stripping the enclosing double-quotes from the parameter will also strip any others.

However, it's easily* solved by supplying the transform as it is (with the single quotes included), and then replacing single quotes with doubles quotes (AFTER first stripping off the beginning and ending double quotes which enclose the parameter.
`set transform=%transform:"=%`
`set transform=%transform:'="%`

_* - if anything can be considered "easy" in DOS batch scripts ;-)_

**Reply by fmw42** [reply]
What do you mean by gradient transform? Please clarify with an example or reference?  I do not know that term. If you mean those created with -fx, you can use double quotes in place of single quotes on Windows for them

**Reply by snibgo** [reply]
On dequoting: I suggest:
`set transform=%~1`
This removes leading and trailing quotes only. See "help set" and the last part of "help for".

**Reply by Traitorsclaw** [reply]
"Gradient Transform" is not my terminology, but a heading that Anthony has given to a final group of more complex transforms near the bottom of the page (referenced in my OP).

**Reply by fmw42** [reply]
As I mentioned, you can substitute double quotes for the single one for -fx and -draw.

**Comment by fmw42** [comment]
Latest version is here https://imagemagick.org/download/binaries/

**Comment by Traitorsclaw** [comment]
Thanks everyone for the comments. An upgrade to the later version solved the issue I had seen ... and a useful side-effect of this thread seems to have been discovering and fixing that repetition bug in v7 as well.
